firestore/bulkwriter.go

```diff
@@ -0,0 +1,319 @@
+// and set are all supported.
+//
+// Only one operation (create, set, update, delete) per document is allowed.
+// BulkWriter cannot promise atomicity: individual writes can fail or succeed
+// independent of each other. Bulkwriter does not apply writes in any set order;
+// thus a document can't have set on it immediately after creation.
+type BulkWriter struct {
+	database        string             // the database as resource name: projects/[PROJECT]/databases/[DATABASE]
+	start           time.Time          // when this BulkWriter was started; used to calculate qps and rate increases
+	vc              *vkit.Client       // internal client
+	maxOpsPerSecond int                // number of requests that can be sent per second
+	docUpdatePaths  map[string]bool    // document paths with corresponding writes in the queue
+	cancel          context.CancelFunc // context to send cancel message
+	bundler         *bundler.Bundler   // handle bundling up writes to Firestore
+	ctx             context.Context    // context for canceling all BulkWriter operations
+	openLock        sync.Mutex         // guards against setting isOpen concurrently
+	isOpen          bool               // flag that the BulkWriter is closed
+}
+
+// newBulkWriter creates a new instance of the BulkWriter. This
+// version of BulkWriter is intended to be used within go routines by the
+// callers.
+func newBulkWriter(ctx context.Context, c *Client, database string) *BulkWriter {
+	// Although typically we shouldn't store Context objects, in this case we
+	// need to pass this Context through to the Bundler
+	ctx, cancel := context.WithCancel(withResourceHeader(ctx, c.path()))
+
+	bw := &BulkWriter{
+		database:        database,
+		start:           time.Now(),
+		vc:              c.c,
+		isOpen:          true,
+		maxOpsPerSecond: defaultStartingMaximumOpsPerSecond,
+		docUpdatePaths:  make(map[string]bool),
+		cancel:          cancel,
+		ctx:             ctx,
+	}
+
+	// can't initialize with struct above; need instance reference to BulkWriter.send()
+	bw.bundler = bundler.NewBundler(BulkWriterJob{}, bw.send)
+	bw.bundler.HandlerLimit = bw.maxOpsPerSecond
+	bw.bundler.BundleCountThreshold = maxBatchSize
+	bw.bundler.BundleByteLimit = 0 // unlimited size
+
+	return bw
+}
+
+// End sends all enqueued writes in parallel and closes the BulkWriter to new requests.
+// After calling End(), calling any additional method automatically returns
+// with an error. This method completes when there are no more pending writes
+// in the queue.
+func (b *BulkWriter) End() {
+	b.Flush()
+	b.openLock.Lock()
+	b.isOpen = false
+	b.openLock.Unlock()
+}
+
+// Flush commits all writes that have been enqueued up to this point in parallel.
+// This method blocks execution.
+func (bw *BulkWriter) Flush() {
+	bw.bundler.Flush()
+}
+
+// IsOpen gets the current open or closed state of the BulkWriter.
+func (bw *BulkWriter) IsOpen() bool {
```

Comment:
Is this method or the idea of being open needed? Only gets set to false when calling End which is a blocking call.

Reply:
I have the same thought. What if `End` set `isOpen` to false and then any subsequent write just returned an error if a client tried to submit a request after closing the writer? You'd be able to remote this check here.

Reply:
Yeah, we need the idea of being open so that callers don't continually attempt to send to a BW.

I'll remove the `IsOpen()` method tho.